File: src/utils/text.py
"""
Text parsing and formatting utilities for structured LLM responses.

All pack_* functions convert structured data → prompt strings.
All parse_* functions convert raw LLM output → structured data.
"""

from __future__ import annotations

from typing import List

from src.config.constants import LLM_TAB_SEPARATOR
from src.schemas.classification import Classification


def attribute_unify(value: str) -> str:
    return value.lower().replace(" ", "_")


# ---------------------------------------------------------------------------
# Classification helpers
# ---------------------------------------------------------------------------

_VALID_SOURCES = frozenset({"code", "profile", "event"})


def pack_classifications_into_string(
    classifications: List[Classification],
) -> str:
    """Serialise a list of classifications into the tab-separated prompt format.

    Example output::
        profile::My name is Alice
        code::Write me a hello-world script
    """
    lines = [
        f"{c['source']}{LLM_TAB_SEPARATOR}{c['query']}"
        for c in classifications
    ]
    return "\n".join(lines)


def parse_raw_response_to_classifications(content: str) -> List[Classification]:
    """Parse the raw LLM response into a list of Classification dicts.

    Expected line format::
        SOURCE::QUERY
    """
    classifications: List[Classification] = []

    for line in content.strip().splitlines():
        line = line.strip()
        if LLM_TAB_SEPARATOR not in line:
            continue

        parts = line.split(LLM_TAB_SEPARATOR, maxsplit=1)

        if len(parts) < 2:
            continue

        source = parts[0].strip().lower()
        query = parts[1].strip()

        if source in _VALID_SOURCES and query:
            classifications.append({"source": source, "query": query})

    return classifications


# ---------------------------------------------------------------------------
# Profile helpers
# ---------------------------------------------------------------------------


def pack_profiles_into_string(facts: list) -> str:
    """Serialise a list of ProfileFact objects into the prompt format.

    Example output::
        work::company::Google
        basic_info::name::Alice

    """
    lines: list[str] = []
    for f in facts:
        topic = f.topic if hasattr(f, "topic") else f["topic"]
        sub = f.sub_topic if hasattr(f, "sub_topic") else f["sub_topic"]
        memo = f.memo if hasattr(f, "memo") else f["memo"]
        lines.append(
            f"{attribute_unify(topic)}{LLM_TAB_SEPARATOR}"
            f"{attribute_unify(sub)}{LLM_TAB_SEPARATOR}{memo.strip()}"
        )
    return "\n".join(lines) if lines else "NONE"


def parse_raw_response_to_profiles(content: str) -> list[dict]:
    """Parse the raw LLM response into a list of profile fact dicts.

    Expected line format (after the ``---`` separator)::
        TOPIC::SUB_TOPIC::MEMO

    Lines before ``---`` are treated as the LLM's "thinking" and ignored.
    """
    facts: list[dict] = []

    # Skip the thinking section (everything before '---')
    if "---" in content:
        content = content.split("---", maxsplit=1)[1]

    for line in content.strip().splitlines():
        line = line.strip()

        if LLM_TAB_SEPARATOR not in line:
            continue

        parts = line.split(LLM_TAB_SEPARATOR)

        if len(parts) >= 3:
            topic = parts[0].strip().lower()
            sub_topic = parts[1].strip().lower()
            memo = LLM_TAB_SEPARATOR.join(parts[2:]).strip()  # rejoin if memo had separator

            if topic and sub_topic and memo:
                facts.append({
                    "topic": topic,
                    "sub_topic": sub_topic,
                    "memo": memo,
                })

    return facts


# ---------------------------------------------------------------------------
# Temporal / Event helpers
# ---------------------------------------------------------------------------


def parse_raw_response_to_event(content: str) -> dict | None:
    """Parse the raw LLM response for temporal event extraction.

    Expected format::

        DATE: MM-DD
        EVENT_NAME: <name>
        YEAR: <year or empty>
        DESC: <description>
        TIME: <time or empty>
        DATE_EXPRESSION: <original date expression>

    Or simply ``NO_EVENT``.

    Returns:
        Dict with event data or *None* if no event was found.
    """
    content = content.strip()

    if "NO_EVENT" in content.upper():
        return None

    event_data: dict = {
        "date": None,
        "event_name": None,
        "year": None,
        "desc": None,
        "time": None,
        "date_expression": None,
    }

    for line in content.splitlines():
        line = line.strip()
        if not line:
            continue

        upper = line.upper()

        if upper.startswith("DATE:") and not upper.startswith("DATE_EXPRESSION:"):
            value = line[5:].strip()
            if value:
                event_data["date"] = value

        elif upper.startswith("EVENT_NAME:"):
            value = line[11:].strip()
            if value:
                event_data["event_name"] = value

        elif upper.startswith("YEAR:"):
            value = line[5:].strip()
            if value:
                try:
                    event_data["year"] = int(value)
                except ValueError:
                    event_data["year"] = value

        elif upper.startswith("DESC:"):
            value = line[5:].strip()
            if value:
                event_data["desc"] = value

        elif upper.startswith("TIME:"):
            value = line[5:].strip()
            if value:
                event_data["time"] = value

        elif upper.startswith("DATE_EXPRESSION:"):
            value = line[16:].strip()
            if value:
                event_data["date_expression"] = value

    # Must have at least a date to be valid
    if not event_data["date"]:
        return None

    return event_data


# ---------------------------------------------------------------------------
# Image helpers
# ---------------------------------------------------------------------------


def parse_raw_response_to_image(content: str) -> dict:
    """Parse the raw LLM response for image analysis.

    Expected format::

        DESCRIPTION: <natural language summary>

        OBSERVATIONS:
        - [category] observation text (confidence: high/medium/low)
        - [category] observation text (confidence: high/medium/low)

    Returns:
        Dict with keys ``description`` (str) and ``observations`` (list of dicts).
        Each observation dict has ``category``, ``description``, and optional ``confidence``.
    """
    content = content.strip()

    result: dict = {
        "description": "",
        "observations": [],
    }

    # --- Extract DESCRIPTION line ---
    for line in content.splitlines():
        stripped = line.strip()
        if stripped.upper().startswith("DESCRIPTION:"):
            result["description"] = stripped[len("DESCRIPTION:"):].strip()
            break

    # --- Extract OBSERVATIONS section ---
    in_observations = False
    for line in content.splitlines():
        stripped = line.strip()

        if stripped.upper().startswith("OBSERVATIONS:"):
            in_observations = True
            continue

        if not in_observations or not stripped.startswith("-"):
            continue

        # Remove the leading "- "
        entry = stripped[1:].strip()

        # Parse "[category] description (confidence: level)"
        category = "other"
        confidence = None
        description = entry

        # Extract category from [brackets]
        if entry.startswith("[") and "]" in entry:
            bracket_end = entry.index("]")
            category = entry[1:bracket_end].strip().lower()
            description = entry[bracket_end + 1:].strip()

        # Extract confidence from (confidence: level)
        if "(confidence:" in description.lower():
            idx = description.lower().index("(confidence:")
            conf_part = description[idx:]
            description = description[:idx].strip()

            conf_part = conf_part.strip("()")
            if ":" in conf_part:
                confidence = conf_part.split(":", maxsplit=1)[1].strip().lower()

        if description:
            result["observations"].append({
                "category": category,
                "description": description,
                "confidence": confidence,
            })

    return result

File: xmem-go/internal/utils/text.go
package utils

import (
	"strconv"
	"strings"
)

const LLMTabSeparator = "::"

var validSources = map[string]struct{}{
	"code":    {},
	"profile": {},
	"event":   {},
	"image":   {},
}

type Classification struct {
	Source string
	Query  string
}

type ProfileFact struct {
	Topic    string
	SubTopic string
	Memo     string
}

type EventData struct {
	Date           string
	EventName      string
	Year           string
	Desc           string
	Time           string
	DateExpression string
}

type ImageObservation struct {
	Category    string
	Description string
	Confidence  string
}

type ImageAnalysisResult struct {
	Description  string
	Observations []ImageObservation
}

func AttributeUnify(value string) string {
	return strings.ToLower(strings.ReplaceAll(value, " ", "_"))
}

func PackClassificationsIntoString(classifications []Classification) string {
	lines := make([]string, 0, len(classifications))
	for _, c := range classifications {
		lines = append(lines, c.Source+LLMTabSeparator+c.Query)
	}
	return strings.Join(lines, "\n")
}

func ParseRawResponseToClassifications(content string) []Classification {
	classifications := []Classification{}
	for _, line := range strings.Split(strings.TrimSpace(content), "\n") {
		line = strings.TrimSpace(line)
		if !strings.Contains(line, LLMTabSeparator) {
			continue
		}
		parts := strings.SplitN(line, LLMTabSeparator, 2)
		if len(parts) < 2 {
			continue
		}
		source := strings.ToLower(strings.TrimSpace(parts[0]))
		query := strings.TrimSpace(parts[1])
		if _, ok := validSources[source]; ok && query != "" {
			classifications = append(classifications, Classification{Source: source, Query: query})
		}
	}
	return classifications
}

func PackProfilesIntoString(facts []ProfileFact) string {
	lines := make([]string, 0, len(facts))
	for _, f := range facts {
		lines = append(lines,
			AttributeUnify(f.Topic)+LLMTabSeparator+
				AttributeUnify(f.SubTopic)+LLMTabSeparator+
				strings.TrimSpace(f.Memo))
	}
	if len(lines) == 0 {
		return "NONE"
	}
	return strings.Join(lines, "\n")
}

func ParseRawResponseToProfiles(content string) []ProfileFact {
	facts := []ProfileFact{}
	if strings.Contains(content, "---") {
		parts := strings.SplitN(content, "---", 2)
		content = parts[1]
	}
	for _, line := range strings.Split(strings.TrimSpace(content), "\n") {
		line = strings.TrimSpace(line)
		if !strings.Contains(line, LLMTabSeparator) {
			continue
		}
		parts := strings.Split(line, LLMTabSeparator)
		if len(parts) >= 3 {
			topic := strings.ToLower(strings.TrimSpace(parts[0]))
			subTopic := strings.ToLower(strings.TrimSpace(parts[1]))
			memo := strings.TrimSpace(strings.Join(parts[2:], LLMTabSeparator))
			if topic != "" && subTopic != "" && memo != "" {
				facts = append(facts, ProfileFact{
					Topic:    topic,
					SubTopic: subTopic,
					Memo:     memo,
				})
			}
		}
	}
	return facts
}

func ParseRawResponseToEvents(content string) []EventData {
	content = strings.TrimSpace(content)
	if strings.Contains(strings.ToUpper(content), "NO_EVENT") {
		return []EventData{}
	}

	rawBlocks := strings.Split(content, "---")
	blocks := make([]string, 0, len(rawBlocks))
	for _, b := range rawBlocks {
		trimmed := strings.TrimSpace(b)
		if trimmed != "" {
			blocks = append(blocks, trimmed)
		}
	}

	events := []EventData{}
	for _, block := range blocks {
		if event, ok := parseSingleEventBlock(block); ok {
			events = append(events, event)
		}
	}
	if len(events) == 0 {
		if event, ok := parseSingleEventBlock(content); ok {
			events = append(events, event)
		}
	}
	return events
}

func parseSingleEventBlock(block string) (EventData, bool) {
	if strings.Contains(strings.ToUpper(block), "NO_EVENT") {
		return EventData{}, false
	}
	eventData := EventData{}
	for _, line := range strings.Split(block, "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		upper := strings.ToUpper(line)

		if strings.HasPrefix(upper, "DATE:") && !strings.HasPrefix(upper, "DATE_EXPRESSION:") {
			value := strings.TrimSpace(line[5:])
			if value != "" {
				eventData.Date = value
			}
		} else if strings.HasPrefix(upper, "EVENT_NAME:") {
			value := strings.TrimSpace(line[11:])
			if value != "" {
				eventData.EventName = value
			}
		} else if strings.HasPrefix(upper, "YEAR:") {
			value := strings.TrimSpace(line[5:])
			if value != "" {
				if _, err := strconv.Atoi(value); err == nil {
					eventData.Year = value
				} else {
					eventData.Year = value
				}
			}
		} else if strings.HasPrefix(upper, "DESC:") {
			value := strings.TrimSpace(line[5:])
			if value != "" {
				eventData.Desc = value
			}
		} else if strings.HasPrefix(upper, "TIME:") {
			value := strings.TrimSpace(line[5:])
			if value != "" {
				eventData.Time = value
			}
		} else if strings.HasPrefix(upper, "DATE_EXPRESSION:") {
			value := strings.TrimSpace(line[16:])
			if value != "" {
				eventData.DateExpression = value
			}
		}
	}
	if eventData.Date == "" {
		return EventData{}, false
	}
	return eventData, true
}

func ParseRawResponseToImage(content string) ImageAnalysisResult {
	content = strings.TrimSpace(content)
	result := ImageAnalysisResult{
		Description:  "",
		Observations: []ImageObservation{},
	}

	for _, line := range strings.Split(content, "\n") {
		stripped := strings.TrimSpace(line)
		if strings.HasPrefix(strings.ToUpper(stripped), "DESCRIPTION:") {
			result.Description = strings.TrimSpace(stripped[len("DESCRIPTION:"):])
			break
		}
	}

	inObservations := false
	for _, line := range strings.Split(content, "\n") {
		stripped := strings.TrimSpace(line)
		if strings.HasPrefix(strings.ToUpper(stripped), "OBSERVATIONS:") {
			inObservations = true
			continue
		}
		if !inObservations || !strings.HasPrefix(stripped, "-") {
			continue
		}
		entry := strings.TrimSpace(stripped[1:])
		category := "other"
		confidence := ""
		description := entry

		if strings.HasPrefix(entry, "[") && strings.Contains(entry, "]") {
			bracketEnd := strings.Index(entry, "]")
			category = strings.ToLower(entry[1:bracketEnd])
			description = strings.TrimSpace(entry[bracketEnd+1:])
		}

		lowerDesc := strings.ToLower(description)
		if strings.Contains(lowerDesc, "(confidence:") {
			idx := strings.Index(lowerDesc, "(confidence:")
			confPart := description[idx:]
			description = strings.TrimSpace(description[:idx])
			confPart = strings.Trim(confPart, "()")
			if strings.Contains(confPart, ":") {
				parts := strings.SplitN(confPart, ":", 2)
				confidence = strings.ToLower(strings.TrimSpace(parts[1]))
			}
		}

		if description != "" {
			result.Observations = append(result.Observations, ImageObservation{
				Category:    category,
				Description: description,
				Confidence:  confidence,
			})
		}
	}

	return result
}
